src/node_mime.cc

```diff
@@ -0,0 +1,369 @@
+#include <string>
+#include "node.h"
+#include "env.h"
+#include "node_mime.h"
+#include "node_internals.h"
+namespace node {
+
+namespace mime {
+
+using v8::Array;
+using v8::Context;
+using v8::Function;
+using v8::FunctionCallbackInfo;
+using v8::FunctionTemplate;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::Local;
+using v8::Object;
+using v8::String;
+using v8::Value;
+
+
+// utilities for
+namespace {
+
+bool is_ascii_upper_alpha(char c) {
+  return c >= '\x41' && c <= '\x5A';
+}
+bool is_ascii_lower_alpha(char c) {
+  return c >= '\x61' && c <= '\x7A';
+}
+bool is_ascii_digit(char c) {
+  return c >= '\x30' && c <= '\x39';
+}
+bool is_ascii_alpha(char c) {
+  return is_ascii_upper_alpha(c) || is_ascii_lower_alpha(c);
+}
+bool is_ascii_alphanumeric(char c) {
+  return is_ascii_digit(c) || is_ascii_alpha(c);
+}
+char to_ascii_lower(char c) {
+  if (!is_ascii_upper_alpha(c)) return c;
+  return c + 0x20;
+}
+bool is_http_token(char c) {
+  return c == '\x21' ||
+      c == '\x23' ||
+      c == '\x24' ||
+      c == '\x25' ||
+      c == '\x26' ||
+      c == '\x27' ||
+      c == '\x2A' ||
+      c == '\x2B' ||
+      c == '\x2D' ||
+      c == '\x2E' ||
+      c == '\x5E' ||
+      c == '\x5F' ||
+      c == '\x60' ||
+      c == '\x7C' ||
+      c == '\x7E' ||
+      is_ascii_alphanumeric(c);
+}
+bool is_ascii_whitespace(char c) {
+  return c == '\x09' ||
+      c == '\x0A' ||
+      c == '\x0C' ||
+      c == '\x0D' ||
+      c == '\x20';
+}
+bool is_http_quoted_string_token(char c) {
+  return c == '\x09' ||
+      (c >= '\x20' && c <= '\x7E') ||
+      (c >= '\x80' && c <= '\xFF');
+}
+
+}   // anonymous namespace
+
+MIME::MIME(std::string src) {
+  size_t left = 0;
+  auto saw_invalid_char = false;
+  size_t first_invalid_char = 0;
+  for (; left < src.size(); left++) {
+    auto c = src[left];
+    if (!is_ascii_whitespace(c)) break;
+  }
+  std::string& type = type_;
+  std::string& subtype = subtype_;
+  auto& parameters = parameters_;
+  size_t right = left;
+  for (; right < src.size(); right++) {
+    auto c = src[right];
+    if (c == '/') {
+      if (left == right || right >= src.size() || saw_invalid_char) {
+        this->flags_ |= MIME_FLAGS_INVALID_TYPE;
+        return;
+      }
+      type = src.substr(left, right - left);
+      break;
+    } else if (!saw_invalid_char && !is_http_token(c)) {
+      saw_invalid_char = true;
+      first_invalid_char = right;
+    }
+  }
+  if (type.size() == 0) {
+    this->flags_ |= MIME_FLAGS_INVALID_TYPE;
+    return;
+  }
+  saw_invalid_char = false;
+  right++;
+  left = right;
```

Comment:
It’s not clear what these variables mean, maybe they could have more expressive names?

Reply:
they are the left index of the string parsing and the right index of the string parsing stuff in the middle is the substring being parsed. don't know of a better name as they change in meaning if you want something about what part of the substring is being processed. it would be very noisy I think to have separate ones for each possible substring.